File: app/src/main/java/com/example/android/securenetworkpinning/RequestManager.java
package com.example.android.securenetworkpinning;


import android.content.Context;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.net.URLConnection;

public class RequestManager {

    private static RequestManager instance;

    public static synchronized RequestManager getInstance(Context context) {
        if (instance == null) {
            instance = new RequestManager(context.getApplicationContext());
        }

        return instance;
    }

    private RequestManager(Context context) {
        //TODO: Create and load the local trust store
    }

    public String makeRequest(String endpoint) throws IOException {
        URL url = new URL(endpoint);
        HttpURLConnection urlConnection = (HttpURLConnection)url.openConnection();

        return parseResponse(urlConnection);
    }

    private String parseResponse(URLConnection connection) throws IOException {
        InputStream in = connection.getInputStream();
        String encoding = connection.getContentEncoding();
        int contentLength = connection.getContentLength();
        if (encoding == null) {
            encoding = "UTF-8";
        }

        byte[] buffer = new byte[16384];

        int length = Math.max(contentLength, 0);
        ByteArrayOutputStream out = new ByteArrayOutputStream(length);

        int read;
        while ((read = in.read(buffer)) != -1) {
            out.write(buffer, 0, read);
        }

        return new String(out.toByteArray(), encoding);
    }
}

File: app/src/main/java/com/example/android/securenetworkpinning/RequestManager.kt
package com.example.android.securenetworkpinning

import java.io.ByteArrayOutputStream
import java.io.IOException
import java.net.HttpURLConnection
import java.net.URL
import java.net.URLConnection
import java.nio.charset.Charset
import kotlin.math.max

object RequestManager {
    @Throws(IOException::class)
    fun makeRequest(endpoint: String): String {
        val url = URL(endpoint)
        val urlConnection = url.openConnection() as HttpURLConnection
        return parseResponse(urlConnection)
    }

    @Throws(IOException::class)
    private fun parseResponse(connection: URLConnection): String {
        val input = connection.getInputStream()
        val encoding = connection.contentEncoding ?: "UTF-8"
        val contentLength = connection.contentLength

        val buffer = ByteArray(16384)
        val length = max(contentLength, 0)
        val out = ByteArrayOutputStream(length)
        var read: Int
        while (input.read(buffer).also { read = it } != -1) {
            out.write(buffer, 0, read)
        }
        return String(out.toByteArray(), Charset.forName(encoding))
    }
}